### PR TITLE
Add feature of material farming in task RogueWorld

### PR DESCRIPTION
**Pull request by Schwarze-Katze**
When `RogueWorld_WeeklyFarming` is enabled, task `RogueWorld` will not stop when weekly points are full, instead farming the boss 33 times (i.e., 11 simulated universe) to obtain trace materials.
`RogueWorld_WeeklyFarmingCount` denotes how many turns of rogue remain to complete this week and will be reset to 11 each Monday 04:00.
Does not handle whether the last rogue run failed due to one who use this feature may have less chance of failure.

#231 #247 #264 #288 #294

Correction: Sorry I was misled by #288. Rogue boss has 100 drop times per week, not 33 times. See [this](https://wiki.biligame.com/sr/%E6%A8%A1%E6%8B%9F%E5%AE%87%E5%AE%99%E6%8E%89%E8%90%BD%E6%9C%BA%E5%88%B6).

### CODE REVIEW

**Review by Schwarze-Katze — COMMENTED** [review]
Should I discard the modification of .gitignore or seperate it into a new PR?

**Review by LmeSzinc — COMMENTED** [review]
> Im not quite sure about the usage of self.config.stored.SimulatedUniverse.is_expired() function.

Check out the echo of war countdown
https://github.com/LmeSzinc/StarRailCopilot/blob/58e054a008013c8a1be869b3014c3ba2a3365df2/module/config/stored/classes.py#L284-L285
https://github.com/LmeSzinc/AzurLaneAutoScript/wiki/6.1.-GUI-Option

> Should I discard the modification of .gitignore or seperate it into a new PR?

yes, and comment that these are electron files

**Comment by foodtooth** [comment]
why adding an extra local variable for this while we can just detect the material result after the elite fight? Extra local variable means redundant logic and conflicts as you may wanna do some dlc runs in-game by yourself

**Comment by Zebartin** [comment]
> why adding an extra local variable for this while we can just detect the material result after the elite fight? Extra local variable means redundant logic and conflicts as you may wanna do some dlc runs in-game by yourself

you are right but is there any *material result* that is easy and reliable to detect? 
- the combat rewards shown at the left side disappear in seconds, easy but not reliable
- detect item quantities or credit, reliable but time-consuming

**Comment by foodtooth** [comment]
> > why adding an extra local variable for this while we can just detect the material result after the elite fight? Extra local variable means redundant logic and conflicts as you may wanna do some dlc runs in-game by yourself
> 
> you are right but is there any _material result_ that is easy and reliable to detect?
> 
> * the combat rewards shown at the left side disappear in seconds, easy but not reliable
> * detect item quantities or credit, reliable but time-consuming

Not necessarily relying on any specific material, simply employing template matching for the multiplication sign and counting the quantity should be reliable enough. any thoughts?

File: tasks/rogue/entry/entry.py [hunk 1]
```diff
@@ -347,11 +348,16 @@ def check_stop_condition(self):
             return
 
         if self.config.stored.SimulatedUniverse.is_expired():
-            # Expired, do rogue
-            pass
+            # Expired, do rogue and reset weekly farming count
+            self.config.RogueWorld_WeeklyFarmingCount = 11
+
```
[the remark this answers is not in the document]
Comment (Schwarze-Katze) [reply tasks/rogue/entry/entry.py:353]: Understood. Thank your correction and I ll fix that later today.

**Comment by Schwarze-Katze** [comment]
I have taken @LmeSzinc 's suggestion and set a farming counter in `self.config.stored.SimulatedUniverse` class.
This counter starts from 100, and decreases each time the elite boss is cleared. When the counter is beyond 0, task `SimulatedUniverse` enters rogue route even though weekly rogue point is full.

I have noted the idea of detecting boss drop list by @foodtooth 's suggestion.
This way has several pros and cons:
- ✔️More stable and energy-saving when user plays rogue both manually and by SRC
- ✔️Avoids adding more and more variables
- ❌Lacks of robustness, the drop list disappears soon thus may be ignored due to lag of simulator
- ❌Lacks of versatility, no other modules can benefit from this feature, thus has a low input-output ratio

I cannot judge which way is far better than the other, so I adopted the former one. Furthermore, consider of maintainability, every variable name in this commit contains `farm` prefix, which is unique in `SimulatedUniverse` module, making it easy to pick out for future improvement.

**Comment by Schwarze-Katze** [comment]
🥲It seems that StoredSimulatedUniverse class will not save the counter to disk, causing the counter to reset when src restarts. I dont know how to save this counter, should I use [this way](https://github.com/LmeSzinc/AzurLaneAutoScript/wiki/6.1.-GUI-Option#%E4%BC%A0%E9%80%92%E7%89%B9%E6%AE%8A%E5%B1%9E%E6%80%A7) or adding the counter value directly to GUI?

Another known issue: if SRC is terminated after clearing elite boss but before exiting elite boss's dungeon(which means just a little walk and one click, will not take much time), farming counter will decrease again at the next time the `RogueWorld` task enters that elite dungeon.

File: module/config/stored/classes.py [hunk 2]
```diff
@@ -205,7 +205,25 @@ class StoredImmersifier(StoredCounter):
 
 
 class StoredSimulatedUniverse(StoredCounter, StoredExpiredAtMonday0400):
-    pass
+    # These variables are used in Rogue Farming feature.
+    
+    # Times of boss drop chance per week. In current version of StarRail, this value is 100.
+    FIXED_DEFAULT = 100
```
Comment (LmeSzinc) [inline module/config/stored/classes.py:211]: Add standalone storage
```python
class StoredSimulatedUniverseElite(StoredCounter, StoredExpiredAtMonday0400):
    pass
```

File: tasks/rogue/route/base.py [hunk 3]
```diff
@@ -170,6 +170,13 @@ def clear_elite(self, *waypoints):
         pass
 
         result = super().clear_enemy(*waypoints)
+
+        # runs when one elite battle finishes (do not handle whether the battle succeeded or failed), and decreases rogue farming count by 1
+        if self.config.RogueWorld_WeeklyFarming and self.config.stored.SimulatedUniverse.farm_not_complete():
+            self.config.stored.SimulatedUniverse.farm_dec()
+            logger.info(
+                f"Cleared elite boss, decrease farming count by 1, now {self.config.stored.SimulatedUniverse.FarmingCounter}")
```
Comment (LmeSzinc) [inline tasks/rogue/route/base.py:178]: check the `result` above, reduce counter if `"enemy" in result`

**Comment by Schwarze-Katze** [comment]
Thanks to @LmeSzinc 's suggestion, I have fixed the problem.
Now `MapControl.clear_enemy(self, *waypoints)` and `MapControl.clear_item(self, *waypoints)` methods will return result array, instead of `None`. I think these two methods' not returning anything is a bug, so I let them to return `self.goto(*waypoints)`‘s result.
Any code who expect these two methods to return None may have unexpected result, but during my review no one uses the return value except mine.

**Pull request merged**